src/LogicTasks/Semantics/Resolve.hs

```diff
@@ -185,6 +179,42 @@ completeGrade ResolutionInst{..} sol = do
         german "Letzter Schritt leitet die leere Klausel ab?"
         english "The last step derives the empty clause?"
 
+    pure ()
+
+  pure ()
+  where
+    checkMapping = correctMapping sol $ baseMapping clauses
+    steps =  replaceAll sol $ baseMapping clauses
+    availLits = unions (map (fromList . literals) clauses)
+    stepLits (c1,c2,r) = toList $ unions $ map (fromList . literals) [c1,c2,r]
+    wrongLitsSteps = filter (not . all (`member` availLits) . stepLits) steps
+    checkEmptyClause = null steps || not (isEmptyClause $ third3 $ last steps)
+    noResolveSteps = filter (\(c1,c2,r) -> maybe True (\x ->
+      fromJust (resolve c1 c2 x) /= r) (resolvableWith c1 c2)) steps
+
+completeGrade :: OutputMonad m => ResolutionInst -> [ResStep] -> LangM m
+completeGrade ResolutionInst{..} sol = do
+    unless showFeedbackOnPartialGrade $ do
+      preventWithHint (not $ null noResolveSteps)
+        (translate $ do
+          german "Alle Schritte sind gültig?"
+          english "All steps are valid?"
+        )
+        (paragraph $ do
+          translate $ do
+            german "Mindestens ein Schritt ist kein gültiger Resolutionsschritt. "
+            english "At least one step is not a valid resolution step. "
+          itemizeM $ map (text . show) noResolveSteps
+          pure ()
+        )
+
+      prevent checkEmptyClause $
+        translate $ do
+          german "Letzter Schritt leitet die leere Klausel ab?"
+          english "The last step derives the empty clause?"
+
+      pure ()
```

Comment:
I agree with the `jscpd` linter that the identical code in `partialGrade` and `completeGrade` here should be factored out.